File: patient-manager/frontend/src/dtos/newUserDto.ts
import type { User } from '@/models/user';
import { formatDate } from '@/utils/formatDate';

export interface NewUserDto {
    firstName: string;
    lastName: string;
    oib: string;
    residence: string;
    birthDate: string; //YYYY-MM-DD
    email: string;
    password: string;
    role: string;
}

export function createNewUserDto(
    user: User,
    password: string
): NewUserDto {
    return {
        firstName: user.firstName,
        lastName: user.lastName,
        oib: user.oib,
        residence: user.residence,
        birthDate: formatDate(user.birthDate),
        email: user.email,
        password: password,
        role: user.role
    };
}

File: patient-manager/dto/newUserDto.go
package dto

import (
	"PatientManager/model"
	"PatientManager/util/cerror"
	"fmt"

	"github.com/google/uuid"
	"go.uber.org/zap"
)

type NewUserDto struct {
	Uuid      string `json:"uuid"`
	FirstName string `json:"firstName" binding:"required,min=2,max=100"`
	LastName  string `json:"lastName" binding:"required,min=2,max=100"`
	OIB       string `json:"oib" binding:"required,len=11"`
	BirthDate string `json:"birthDate" binding:"required,datetime=2006-01-02"`
	Email     string `json:"email" binding:"required,email"`
	Password  string `json:"password"`
	Role      string `json:"role" binding:"required,oneof=doctor patient superadmin"`
}

// ToModel create a model from a dto
func (dto *NewUserDto) ToModel() (*model.User, error) {
	role, err := model.StoUserRole(dto.Role)
	if err != nil {
		zap.S().Error("Failed to parse role = %+v, err = %+v", dto.Role, err)
		return nil, cerror.ErrUnknownRole
	}
	if dto.Uuid != "" {
		_, err := uuid.Parse(dto.Uuid)
		if err != nil {
			zap.S().Errorf("Failed to parse uuid = %s, err = %+v", dto.Uuid, err)
			return nil, cerror.ErrBadUuid
		}
	}

	return &model.User{
		Uuid:      uuid.New(),
		FirstName: dto.FirstName,
		LastName:  dto.LastName,
		Email:     dto.Email,
		Role:      role,
		OIB:       dto.OIB,
	}, nil
}

// FromModel returns a dto from model struct
func (dto *NewUserDto) FromModel(m *model.User) *NewUserDto {
	dto = &NewUserDto{
		Uuid:      m.Uuid.String(),
		FirstName: m.FirstName,
		LastName:  m.LastName,
		Email:     m.Email,
		Role:      fmt.Sprint(m.Role),
	}
	return dto
}
